subiquity/tests/api/test_api.py

```diff
@@ -0,0 +1,521 @@
+            'toggle': None
+        }
+        await self.post('/keyboard', keyboard)
+        await self.post('/source', source_id='ubuntu-server')
+        await self.post('/network')
+        await self.post('/proxy', '')
+        await self.post('/mirror', 'http://us.archive.ubuntu.com/ubuntu')
+        resp = await self.get('/storage/guided')
+        disk_id = resp['disks'][0]['id']
+        choice = {"disk_id": disk_id}
+        await self.post('/storage/v2/guided', choice=choice)
+        await self.post('/storage/v2')
+        await self.get('/meta/status', cur='WAITING')
+        await self.post('/meta/confirm', tty='/dev/tty1')
+        await self.get('/meta/status', cur='NEEDS_CONFIRMATION')
+        identity = {
+            'realname': 'ubuntu',
+            'username': 'ubuntu',
+            'hostname': 'ubuntu-server',
+            'crypted_password': '$6$exDY1mhS4KUYCE/2$zmn9ToZwTKLhCw.b4/'
+                                + 'b.ZRTIZM30JZ4QrOQ2aOXJ8yk96xpcCof0kx'
+                                + 'KwuX1kqLG/ygbJ1f8wxED22bTL4F46P0'
+        }
+        await self.post('/identity', identity)
+        ssh = {
+            'install_server': False,
+            'allow_pw': False,
+            'authorized_keys': []
+        }
+        await self.post('/ssh', ssh)
+        await self.post('/snaplist', [])
+        for state in 'RUNNING', 'POST_WAIT', 'POST_RUNNING', 'UU_RUNNING':
+            await self.get('/meta/status', cur=state)
+
+    @timeout(5)
+    async def test_default(self):
+        choice = {'disk_id': 'disk-sda'}
+        resp = await self.post('/storage/guided', choice=choice)
+        self.assertEqual(7, len(resp['config']))
+        self.assertEqual('/dev/sda', resp['config'][0]['path'])
+
+    @timeout(5)
+    async def test_guided_v2(self):
+        choice = {'disk_id': 'disk-sda'}
+        resp = await self.post('/storage/v2/guided', choice=choice)
+        self.assertEqual(1, len(resp['disks']))
+        self.assertEqual('disk-sda', resp['disks'][0]['id'])
+
+    @timeout(5)
+    async def test_v2_add_boot_partition(self):
+        self.maxDiff = None
+        disk_id = 'disk-sda'
+
+        data = {
+            'disk_id': disk_id,
+            'partition': {
+                'format': 'ext4',
+                'mount': '/',
+            }
+        }
+        simple_add = await self.post('/storage/v2/add_partition', data)
+        self.assertEqual(2, len(simple_add['disks'][0]['partitions']))
+
+        await self.post('/storage/v2/reset')
+
+        await self.post('/storage/v2/add_boot_partition', disk_id=disk_id)
+        manual_add = await self.post('/storage/v2/add_partition', data)
+        self.assertEqual(simple_add, manual_add)
+
+
+class TestWin10(TestAPI):
+    machine_config = 'examples/win10.json'
+
+    @timeout(5)
+    async def test_bitlocker(self):
+        resp = await self.get('/storage/has_bitlocker')
+        self.assertEqual('BitLocker', resp[0]['partitions'][2]['format'])
```

Comment:
Uh, OK this test is, let's say, not clearly correct to the first time reader. Let me dump some reading on you! First a talk I gave ages and ages ago and wrote up into a blog post slightly less ages ago:

https://medium.com/@mwhudson/a-patronizing-guide-to-what-makes-a-good-unit-test-fe895951ce35

And another blog post on the same kind of topic: https://thoughtbot.com/blog/mystery-guest

In some ways, this sort of code (and also probert and curtin) are hard to write totally transparent unit tests for, because while we _could_ write a helper function that constructed a machine config with appropriate fields and then test the server responded appropriately, at the end of the day we really care about how the server responds to an actual probert report of a bitlocker partition, not this codes understanding of what such a report looks like.

So I guess what I think this test should do is assert that there is something about the response that indicates bitlocker is in use, but not assert that it's the 2nd partition of the 0th disk, or whatever. Does that make sense?

Reply:
That particular API returns a (possibly empty) list of disks that are known to have 1 or more Bitlockered partitions on them.  I've adjusted it to just do a simple len check.  
I'll give the links some thought in the coming days.